rfcs/rfc-113.md

```diff
@@ -0,0 +1,341 @@
+
+Assigning parts of the incoming data to variables will allow rule authors to compose simple to read rules:
+```
+%encryption = BucketEncryption.ServerSideEncryptionConfiguration.*.ServerSideEncryptionByDefault
+%encryption.property == "VALUE"
+$encryption.otherProperty IN ["ALLOWED", "VALUES"]
+```
+
+ 
+# Proposal: Enhanced Property Access and Filtering for Guard
+
+The proposed enhancements below provide customers with a simple language that solves many of the rule modelling issues of the current language.
+The language is general enough to work with any structured payload (like JSON) to evaluate a set of rules against them 
+to help audit security, compliance or overall system assessment for the customer. The filtering and property access enhancements allow for a more in depth rule authoring process that gives customers full control over what to check in a given CloudFormation template or any structured json document.
+
+Note that some of the examples below build off the block evaluations proposed in [#111](https://github.com/aws-cloudformation/cloudformation-guard/pull/111).
+## Proposed enhancements
+
+### Non-resource type scoped Property Access/Support
+
+The non-type scoped property access allows access of any portion of incoming structured data (not just CloudFormation Templates). In the previous iteration of the language, rule authors could only check values of a resource whose type was specified in the rule. Now, one could access any values in the incoming template.
+
+For example, given the CloudFormation template:
+```
+{
+    "Parameters": {
+       "BucketName" : {
+            "Type" : "String",
+            "Default" : "defaultBucket",
+            "AllowedValues" : ["mybucket", "prodbucket", "testbucket"],
+            "Description" : "Valid values for buckets created"
+       }
+    }
+    "Resources": {
+        "MyBucket": {
+            "Type": "AWS::S3::Bucket"
+            "Properties": {
+                "BucketName": {
+                    "Ref": "BucketName"
+                }
+            }
+            "DeletionPolicy": "Retain",
+            "UpdateReplacePolicy": "Retain"
+        }
+    }
+}
+```
+
+With the proposed property access and clause evaluation, one can access any structured data in the input, not only resource type properties. Example ruleset:
+
+```
+Parameters.BucketName EXISTS # ensure a certain parameter exists
+Parameters.BucketName.Type == "String" # conforms to a certain type
+Parameters.BucketName.AllowedValues.* == /\w*bucket\w*/ # make sure values are expected
+
+Resources.*[Type == "AWS::S3::Bucket"].Properties.BucketName == { "Ref": "BucketName"} # ensure bucket name is in allowed values of parameters
```

Comment:
similar to https://github.com/aws-cloudformation/cloudformation-guard/pull/111#discussion_r560576236, I'd really rather not have to specify `Resources` and `Properties` in almost every rule, so I'd want something like the block-syntax if they're both going to be required going forward

Reply:
yeah this was done more to show the functionality of the changes without mixing the two proposals. ideally if these are both merged you could choose how to write these checks: using blocks or some sort of filtering on the top level keys like this